### PR TITLE
Fix "may be used uninitialized" warnings.

### LINKED ISSUE #816: maybe uninitialized vars in tables.c
Building on Ubuntu gives me this:

```
/home/bhaller/Desktop/SLiM/treerec/tskit/tables.c: In function ‘tsk_table_collection_add_and_remap_node’:
/home/bhaller/Desktop/SLiM/treerec/tskit/tables.c:8425:19: warning: ‘pop.metadata_length’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 8425 |             ret = tsk_population_table_add_row(
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 8426 |                 &self->populations, pop.metadata, pop.metadata_length);
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bhaller/Desktop/SLiM/treerec/tskit/tables.c:8425:19: warning: ‘pop.metadata’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/bhaller/Desktop/SLiM/treerec/tskit/tables.c:8408:19: warning: ‘ind.metadata_length’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 8408 |             ret = tsk_individual_table_add_row(&self->individuals, ind.flags,
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 8409 |                 ind.location, ind.location_length, ind.metadata, ind.metadata_length);
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bhaller/Desktop/SLiM/treerec/tskit/tables.c:8408:19: warning: ‘ind.metadata’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/bhaller/Desktop/SLiM/treerec/tskit/tables.c:8408:19: warning: ‘ind.location_length’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/bhaller/Desktop/SLiM/treerec/tskit/tables.c:8408:19: warning: ‘ind.location’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/bhaller/Desktop/SLiM/treerec/tskit/tables.c:8408:19: warning: ‘ind.flags’ may be used uninitialized in this function [-Wmaybe-uninitialized]
```

I've been ignoring these warnings because I thought they were spurious – the compiler just not understanding that `tsk_individual_table_get_row()` and `tsk_population_table_get_row()` fill in the entries in the object (`tsk_individual_t` or `tsk_population_t`) that gets passed in by pointer.  But now, looking at the code, I think maybe the compiler is actually being smart.  The problem is that `tsk_individual_table_get_row()` and `tsk_population_table_get_row()` check for an out-of-range index and return an error in that case, *before* they fill in any of the fields in the object passed in by pointer, so there's a code path where the object passed in by pointer is indeed not initialized.  That wouldn't raise a red flag for the compiler, EXCEPT that the caller doesn't check the return value from those two calls.  So if the out-of-range error occurs, it won't be caught, and the code will proceed to use uninitialized values from the objects.  So, compiler warnings FTW.

### PR DESCRIPTION
**Pull request by grahamgower**
Closes #816.

### CODE REVIEW

**Comment by petrelharp** [comment]
Oh, good catch. Thanks, and looks good.

**Comment by jeromekelleher** [comment]
Well, IIRC we left these checks out because the ``get_x`` methods can't fail - we're guaranteed to only call with valid IDs within these loops. So, the compiler is being overly paranoid here. I guess this is the most future-proof way to go though, even if it means we drop some test coverage.

**Review by jeromekelleher — APPROVED** [review]

**Comment by jeromekelleher** [comment]
@benjeffery - do you think it's worth having ``static void`` versions of these entity getter functions which don't do range checking? It's tedious having to check return values to keep the compiler happy, and it might be more readable too. The functions could be inlined too, so we'd speed some inner loops up a bit probably.

**Comment by jeromekelleher** [comment]
Thinking about this more, I think we should add these non-checking getters. I'll tack on a commit doing this when I get a chance.

**Comment by benjeffery** [comment]
> @benjeffery - do you think it's worth having `static void` versions of these entity getter functions which don't do range checking? It's tedious having to check return values to keep the compiler happy, and it might be more readable too. The functions could be inlined too, so we'd speed some inner loops up a bit probably.

Yes, the `int`, range-checked versions can then call the `static void` versions to avoid duplication. We should name the `static_void` versions something like `tsk_individual_table_get_row_no_check` so that it is obvious on code review that we are doing something we need to be careful about.

**Comment by jeromekelleher** [comment]
@benjeffery - I've pushed an update here with ``x_table_get_row_unsafe``, and used it in any situation in which the library itself guarantees that the references are good. This hopefully clarifies the code a bit, and should bump up our test coverage a bit.

File: c/tskit/tables.c [hunk 1]
```diff
@@ -8421,7 +8453,8 @@ tsk_table_collection_add_and_remap_node(tsk_table_collection_t *self,
             population_map[node.population] = node.population;
         }
         if (population_map[node.population] == TSK_NULL) {
-            tsk_population_table_get_row(&other->populations, node.population, &pop);
+            tsk_population_table_get_row_unsafe(
+                &other->populations, node.population, &pop);
```
Comment (benjeffery) [inline c/tskit/tables.c:8457]: Same here, how do we know `other->nodes` has valid individuals and populations? I followed it back a bit but wasn't obvious to me.

**Review by benjeffery — COMMENTED** [review]
Looks great, love the name.
Just two places where I'd like to know how we know they are safe?

**Comment by benjeffery** [comment]
@jeromekelleher not sure you'll get pinged by my review, so manually pinging!

File: c/tskit/tables.c [hunk 2]
```diff
@@ -8397,14 +8425,18 @@ tsk_table_collection_add_and_remap_node(tsk_table_collection_t *self,
...
     ret = tsk_node_table_get_row(&other->nodes, node_id, &node);
     if (ret < 0) {
         goto out;
     }
     new_ind = TSK_NULL;
     if (node.individual != TSK_NULL) {
         if (individual_map[node.individual] == TSK_NULL) {
-            tsk_individual_table_get_row(&other->individuals, node.individual, &ind);
+            tsk_individual_table_get_row_unsafe(
+                &other->individuals, node.individual, &ind);
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply c/tskit/tables.c]: I think a good rule of thumb will be "if there's any doubt, there is no doubt" - so I've reverted to using the safe versions.

[the remark this answers is not in the document]
Comment (grahamgower) [reply c/tskit/tables.c]: ... and then pull your hair out trying to come up with a test case :)

**Comment by jeromekelleher** [comment]
Thanks @benjeffery - I've reverted to the safe versions in the slightly iffy ones you pointed out. Even if they're safe now, the function might get used in an unexpected way in the future, so best to be on the safe side.

**Review by benjeffery — APPROVED** [review]
Looks good - I see we are getting codecov comments in the diff now, cool!

**Comment by jeromekelleher** [comment]
> I see we are getting codecov comments in the diff now, cool!

Yeah, it's great isn't it?

**Pull request merged**